setup.sh

```diff
@@ -1,45 +1,82 @@
-  sudo apt-get -yy install apt-transport-https ca-certificates curl software-properties-common pwgen gnupg
+  apt-get -qqy update
+  DEBIAN_FRONTEND=noninteractive apt-get -qqy -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" upgrade
+  apt-get -yy install apt-transport-https ca-certificates curl software-properties-common pwgen gnupg
+
+  # Add Docker GPG signing key
   if [ ! -f "/etc/apt/keyrings/docker.gpg" ]; then
-    sudo install -m 0755 -d /etc/apt/keyrings
-    curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
-    sudo chmod a+r /etc/apt/keyrings/docker.gpg
+    install -m 0755 -d /etc/apt/keyrings
+    curl -fsSL https://download.docker.com/linux/debian/gpg | gpg --dearmor -o /etc/apt/keyrings/docker.gpg
+    chmod a+r /etc/apt/keyrings/docker.gpg
   fi
-  echo \
-    "deb [arch=""$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
-    ""$(. /etc/os-release && echo "$VERSION_CODENAME") stable" |
-    sudo tee /etc/apt/sources.list.d/docker.list >/dev/null
-  sudo apt-get update && sudo apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
 
-  # Install Docker Compose
-  sudo ln -sfv /usr/libexec/docker/cli-plugins/docker-compose /usr/local/bin/docker-compose
+  # Add Docker download repository to apt
+  cat <<EOF >/etc/apt/sources.list.d/docker.list
+deb [arch=""$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian $(. /etc/os-release && echo "$VERSION_CODENAME") stable
+EOF
+  apt-get update && apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
+}
+
+install_docker_ubuntu() {
+  echo "** Installing Docker (Ubuntu) **"
+
+  export DEBIAN_FRONTEND=noninteractive
+  apt-get -qqy update
+  DEBIAN_FRONTEND=noninteractive sudo -E apt-get -qqy -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" upgrade
+  apt-get -yy install apt-transport-https ca-certificates curl software-properties-common pwgen gnupg
+
+  # Add Docker GPG signing key
+  if [ ! -f "/etc/apt/keyrings/docker.gpg" ]; then
+    install -m 0755 -d /etc/apt/keyrings
+    curl -fsSL https://download.docker.com/linux/ubuntu/gpg | gpg --dearmor -o /etc/apt/keyrings/docker.gpg
+    chmod a+r /etc/apt/keyrings/docker.gpg
+  fi
 
-  # Allow current user to run Docker commands
-  sudo usermod -aG docker "$USER"
+  # Add Docker download repository to apt
+  cat <<EOF >/etc/apt/sources.list.d/docker.list
+deb [arch=""$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu $(. /etc/os-release && echo "$VERSION_CODENAME") stable
+EOF
+  apt-get update && apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
```

Comment:
After installing should we set the `docker` group so users can use `docker` without `sudo`?


Like:
```sh
groupadd docker || true # don't fail to add if it already exists
usermod -aG docker $USER
newgrp docker
```

https://docs.docker.com/engine/install/linux-postinstall

Reply:
Normally I'd say yes, but in this one specific case the script is designed to be run as the root user.  Either directly, or from sudo.  So it would effectively be a no-op if we added it back in. :wink:

The previous script, which embedded sudo's throughout it, was able to better add the user to the `docker` group as you suggest.

Am open to ideas if there's some way of making it work... without adding a bunch of sudo's back into the script. :smile:

Reply:
I guess we could add some kind of post install instruction about it, so the user is able to add themself to the group if appropriate.

Reply:
Been thinking this over, and for now lets also skip adding a post install text piece about this.  The post-install will start to look like a wall of text otherwise, which will mean people won't read it.

If it turns out that was a dumb idea on my part and we should add it after all, then we can do so then. :smile: